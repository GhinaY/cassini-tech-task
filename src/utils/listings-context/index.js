import { createContext, useCallback, useContext, useState } from 'react';

export const ListingsContext = createContext(null);

export const useListingsContext = () => useContext(ListingsContext);

function ListingsContextProvider({children}) {
  const [listings, setListings] = useState({});

  const setListingsFromArray = useCallback((newValuesArray) => {
    const newValuesObject = newValuesArray.reduce((acc, listing) => ({ ...acc, [listing.id]: listing}), {});
    setListings(newValuesObject)
  }, []);

  const updateSingleListing = useCallback((newListingValue) => {
    setListings({
      ...listings,
      [newListingValue.id]: newListingValue,
    })
  }, [listings]);

  return (
     <ListingsContext.Provider value={{listings, setListingsFromArray, updateSingleListing}}>
       {children}
     </ListingsContext.Provider>
   );
};

export default ListingsContextProvider;
