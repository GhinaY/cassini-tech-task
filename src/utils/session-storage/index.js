export function updateStorageItem(key, value) {
  value ? sessionStorage.setItem(key, value) : sessionStorage.removeItem(key);
};

export function getStorageItem(key) {
  return sessionStorage.getItem(key);
}