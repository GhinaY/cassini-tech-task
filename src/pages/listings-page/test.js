import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { ListingsContext } from '../../utils/listings-context';
import mockListingsData from '../../utils/test-utils/mock-listings-data';
import ListingsPage from '.';

jest.mock('../../components/product-card', () => ({
  __esModule: true,
  default: ({id}) => (<div>{id}</div>)
}));

jest.mock('../../components/category-filter', () => ({
  __esModule: true,
  default: () => (<div>category filter</div>)
}));

const mockSetListingsFromArray = jest.fn();

function renderListingsPage(props) {
  return (
    render(
      <ListingsContext.Provider value={{
        listings: mockListingsData,
        setListingsFromArray: mockSetListingsFromArray
      }}>
        <MemoryRouter>
          <ListingsPage {...props}/>
        </MemoryRouter>
      </ListingsContext.Provider>
    )
  );
};

describe('ListingsPage', () => {
  beforeEach(() => {
    jest.spyOn(global, 'fetch').mockImplementation(() => Promise.resolve({
      ok: true,
      status: 200,
      json: async () => []
    }))
  });

  test('renders page title', () => {
    const { asFragment } = renderListingsPage();

    expect(asFragment()).toMatchSnapshot();
  });
});

// Definitely needs a lot more tests, including functional ones, 
// but in the interests of time I'll skip them for now.
