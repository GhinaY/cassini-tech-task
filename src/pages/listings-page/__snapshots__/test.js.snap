// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListingsPage renders page title 1`] = `
<DocumentFragment>
  <div
    class="ListingsPageContainer"
  >
    <header
      class="Header"
    >
      <h1
        class="ListingsTitle"
      >
        All Products
      </h1>
      <div
        class="ControlsBar"
      >
        <div>
          category filter
        </div>
        <button
          class="Button"
        >
          Create
        </button>
      </div>
    </header>
    <div
      class="ListingsGrid"
    >
      <div>
        1
      </div>
      <div>
        2
      </div>
      <div>
        3
      </div>
      <div>
        4
      </div>
      <div>
        5
      </div>
      <div>
        6
      </div>
      <div>
        7
      </div>
      <div>
        8
      </div>
      <div>
        9
      </div>
      <div>
        10
      </div>
    </div>
    <button
      class="Button"
    >
      Load more
    </button>
  </div>
</DocumentFragment>
`;
